metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1123,29 +1165,104 @@ private void claim(int epoch) {
             // Prepend the activate event. It is important that this event go at the beginning
             // of the queue rather than the end (hence prepend rather than append). It's also
             // important not to use prepend for anything else, to preserve the ordering here.
-            queue.prepend(new ControllerWriteEvent<>("completeActivation[" + epoch + "]",
-                    new CompleteActivationEvent()));
+            ControllerWriteEvent<Void> activationEvent = new ControllerWriteEvent<>("completeActivation[" + epoch + "]",
+                    new CompleteActivationEvent(),
+                    EnumSet.of(DOES_NOT_UPDATE_QUEUE_TIME, RUNS_IN_PREMIGRATION));
+            activationEvent.future.whenComplete((__, t) -> {
+                if (t != null) {
+                    fatalFaultHandler.handleFault("exception while activating controller", t);
```

Comment:
Yeah, this is fair, I guess. We should invoke the fault handler if we have a failure in the activation event. Although we'll already catch non-`ApiException`. But still. A good check.

Reply:
In `handleEventException`, we renounce the controller leadership when we see a non-ApiException from a ControllerWriteEvent. As far as I can tell, this does not hit the fault handler.

`KafkaEventQueue#run` traps exceptions and calls `Event#handleException`. In `ControllerEvent#handleException`, we call `QuorumController#handleEventException` which just does logging and resignation. `handleEventException` wraps the exception with UnknownServerException, but I don't see where we do anything with that in ControllerEvent.

Reply:
Yes, I recall now. We always hit the fault handler if `replay` fails, but other issues just trigger a resignation and failover. Since this is a bit special it makes sense to specifically invoke the fault handler.

Should we use `exceptionally` here rather than `whenComplete`?